Index: serdisplib-2.01/src/serdisp_specific_framebuffer.c
===================================================================
--- serdisplib-2.01.orig/src/serdisp_specific_framebuffer.c
+++ serdisplib-2.01/src/serdisp_specific_framebuffer.c
@@ -605,12 +605,12 @@ void serdisp_framebuffer_setsdpixel (ser
 
   switch(dd->depth) {
     case 16:
-      if ( dd->scrbuf[idx+1] != ((sdcol & 0xFF00) >> 8) ) {
-        dd->scrbuf[idx+1] = ((sdcol & 0xFF00) >> 8);
+      if ( dd->scrbuf[idx] != ((sdcol & 0xFF00) >> 8) ) {
+        dd->scrbuf[idx] = ((sdcol & 0xFF00) >> 8);
         changed = 1;
       }
-      if ( dd->scrbuf[idx] != (sdcol & 0xFF) ) {
-        dd->scrbuf[idx] = (sdcol & 0x00FF);
+      if ( dd->scrbuf[idx+1] != (sdcol & 0xFF) ) {
+        dd->scrbuf[idx+1] = (sdcol & 0x00FF);
         changed = 1;
       }
       break;
