--- old/drivers/usb/storage/scsiglue.c	2011-03-17 17:44:57.000000000 +0800
+++ new/drivers/usb/storage/scsiglue.c	2011-03-17 17:44:00.000000000 +0800
@@ -541,7 +541,11 @@
 	.sg_tablesize =			SG_ALL,
 
 	/* limit the total size of a transfer to 120 KB */
-	.max_sectors =                  240,
+	#ifdef CONFIG_AMAZON_SE
+	.max_sectors =                  64,
+	#else
+	.max_sectors =                  256,
+	#endif
 
 	/* merge commands... this seems to help performance, but
 	 * periodically someone should test to see which setting is more
