--- a/drivers/net/ethernet/lantiq_xrx200.c
+++ b/drivers/net/ethernet/lantiq_xrx200.c
@@ -758,12 +758,25 @@ static int xrx200_get_port_attr(struct s
 // attributes
 static struct switch_attr xrx200sw_globals[] = {
 	{
+		XRX200_GLOBAL_REGATTR(XRX200_PCE_PMAP_2_DMCPMAP),
+		.name = "dmcpmap",
+		.description = "Default Multicast/Broadcast Port Map",
+		.max = 0x7F,
+	},
+	{
+		XRX200_GLOBAL_REGATTR(XRX200_PCE_PMAP_3_UUCMAP),
+		.name = "uucmap",
+		.description = "Default Unknown Unicast Port Map",
+		.max = 0x7F,
+	},
+	{
 		.type = SWITCH_TYPE_INT,
 		.set = xrx200_set_vlan_mode_enable,
 		.get = xrx200_get_vlan_mode_enable,
 		.name = "enable_vlan",
 		.description = "Enable VLAN mode",
-		.max = 1},
+		.max = 1
+	},
 };
 
 static struct switch_attr xrx200sw_port[] = {
@@ -1524,9 +1537,13 @@ static void xrx200_hw_init(struct xrx200
 	xrx200_pci_microcode();
 
 	/* Default unknown Broadcat/Multicast/Unicast port maps */
-	ltq_switch_w32(0x40, PCE_PMAP1);
-	ltq_switch_w32(0x40, PCE_PMAP2);
-	ltq_switch_w32(0x40, PCE_PMAP3);
+	ltq_switch_w32(0x7F, PCE_PMAP1);
+
+	/* Default unknown Multicast/Broadcast port map (DMCPMAP) (unknown multicast, broadcast port map) */
+	ltq_switch_w32(0x7F, PCE_PMAP2);
+
+	/* Default unknown Unicast port map (UUCMAP) (unknown unicast port map) */
+	ltq_switch_w32(0x7F, PCE_PMAP3);
 
 	/* RMON Counter Enable for all physical ports */
 	for (i = 0; i < 7; i++)
--- a/drivers/net/ethernet/lantiq_xrx200_sw.h
+++ b/drivers/net/ethernet/lantiq_xrx200_sw.h
@@ -277,12 +277,12 @@ enum {
 //	XRX200_PCE_AGE_0_EXP,          /* Aging Counter Exponent Value  */
 //	XRX200_PCE_AGE_1,              /* Aging Counter ConfigurationRegister 1 */
 //	XRX200_PCE_AGE_1_MANT,         /* Aging Counter Mantissa Value  */
-//	XRX200_PCE_PMAP_1,             /* Port Map Register 1 */
-//	XRX200_PCE_PMAP_1_MPMAP,       /* Monitoring Port Map */
-//	XRX200_PCE_PMAP_2,             /* Port Map Register 2 */
-//	XRX200_PCE_PMAP_2_DMCPMAP,     /* Default Multicast Port Map */
-//	XRX200_PCE_PMAP_3,             /* Port Map Register 3 */
-//	XRX200_PCE_PMAP_3_UUCMAP,      /* Default Unknown Unicast Port Map */
+	XRX200_PCE_PMAP_1,             /* Port Map Register 1 */
+	XRX200_PCE_PMAP_1_MPMAP,       /* Monitoring Port Map */
+	XRX200_PCE_PMAP_2,             /* Port Map Register 2 */
+	XRX200_PCE_PMAP_2_DMCPMAP,     /* Default Multicast Port Map */
+	XRX200_PCE_PMAP_3,             /* Port Map Register 3 */
+	XRX200_PCE_PMAP_3_UUCMAP,      /* Default Unknown Unicast Port Map */
 //	XRX200_PCE_GCTRL_0,            /* PCE Global Control Register0 */
 //	XRX200_PCE_GCTRL_0_IGMP,       /* IGMP Mode Selection */
 	XRX200_PCE_GCTRL_0_VLAN,       /* VLAN-aware Switching */
@@ -933,12 +933,12 @@ struct xrx200sw_reg {
 //	{0x1144,	 0,	 4,	0x00}, /* XRX200_PCE_AGE_0_EXP           Aging Counter Exponent Value  */
 //	{0x1148,	 0,	16,	0x00}, /* XRX200_PCE_AGE_1               Aging Counter ConfigurationRegister 1 */
 //	{0x1148,	 0,	16,	0x00}, /* XRX200_PCE_AGE_1_MANT          Aging Counter Mantissa Value  */
-//	{0x114C,	 0,	16,	0x00}, /* XRX200_PCE_PMAP_1              Port Map Register 1 */
-//	{0x114C,	 0,	16,	0x00}, /* XRX200_PCE_PMAP_1_MPMAP        Monitoring Port Map */
-//	{0x1150,	 0,	16,	0x00}, /* XRX200_PCE_PMAP_2              Port Map Register 2 */
-//	{0x1150,	 0,	16,	0x00}, /* XRX200_PCE_PMAP_2_DMCPMAP      Default Multicast Port Map */
-//	{0x1154,	 0,	16,	0x00}, /* XRX200_PCE_PMAP_3              Port Map Register 3 */
-//	{0x1154,	 0,	16,	0x00}, /* XRX200_PCE_PMAP_3_UUCMAP       Default Unknown Unicast Port Map */
+	{0x114C,	 0,	16,	0x00}, /* XRX200_PCE_PMAP_1              Port Map Register 1 */
+	{0x114C,	 0,	16,	0x00}, /* XRX200_PCE_PMAP_1_MPMAP        Monitoring Port Map */
+	{0x1150,	 0,	16,	0x00}, /* XRX200_PCE_PMAP_2              Port Map Register 2 */
+	{0x1150,	 0,	16,	0x00}, /* XRX200_PCE_PMAP_2_DMCPMAP      Default Multicast Port Map */
+	{0x1154,	 0,	16,	0x00}, /* XRX200_PCE_PMAP_3              Port Map Register 3 */
+	{0x1154,	 0,	16,	0x00}, /* XRX200_PCE_PMAP_3_UUCMAP       Default Unknown Unicast Port Map */
 //	{0x1158,	 0,	16,	0x00}, /* XRX200_PCE_GCTRL_0             PCE Global Control Register0 */
 //	{0x1158,	15,	 1,	0x00}, /* XRX200_PCE_GCTRL_0_IGMP        IGMP Mode Selection */
 	{0x1158,	14,	 1,	0x00}, /* XRX200_PCE_GCTRL_0_VLAN        VLAN-aware Switching */
