--- a/drivers/net/ethernet/lantiq_xrx200.c
+++ b/drivers/net/ethernet/lantiq_xrx200.c
@@ -38,6 +38,7 @@
 #define SW_POLLING
 #define SW_ROUTING
 #undef SW_ROUTING
+#define LINK_INITIAL_UP
 
 #define XRX200_MAX_DEV		2
 /*
@@ -1365,6 +1366,10 @@ static int xrx200_mdio_probe(struct net_
 		return PTR_ERR(phydev);
 	}
 
+#ifdef LINK_INITIAL_UP
+	netif_carrier_on(dev);
+#endif
+
 	phydev->supported &= (SUPPORTED_10baseT_Half
 			| SUPPORTED_10baseT_Full
 			| SUPPORTED_100baseT_Half
