--- a/drivers/net/ethernet/lantiq_xrx200.c
+++ b/drivers/net/ethernet/lantiq_xrx200.c
@@ -426,48 +426,6 @@ static int xrx200_pce_table_entry_write(
 	return 0;
 }
 
-static void xrx200sw_fixup_pvids(void)
-{
-	int index, p, portmap, untagged;
-	struct xrx200_pce_table_entry tem;
-	struct xrx200_pce_table_entry tev;
-
-	portmap = 0;
-	for (p = 0; p < XRX200_MAX_PORT; p++)
-		portmap |= BIT(p);
-
-	tem.table = XRX200_PCE_VLANMAP_IDX;
-	tev.table = XRX200_PCE_ACTVLAN_IDX;
-
-	for (index = XRX200_MAX_VLAN; index-- > 0;)
-	{
-		tev.index = index;
-		xrx200_pce_table_entry_read(&tev);
-
-		if (tev.valid == 0)
-			continue;
-
-		tem.index = index;
-		xrx200_pce_table_entry_read(&tem);
-
-		if (tem.val[0] == 0)
-			continue;
-
-		untagged = portmap & (tem.val[1] ^ tem.val[2]);
-
-		for (p = 0; p < XRX200_MAX_PORT; p++)
-			if (untagged & BIT(p))
-			{
-				portmap &= ~BIT(p);
-				xrx200sw_write_x(index, XRX200_PCE_DEFPVID_PVID, p);
-			}
-
-		for (p = 0; p < XRX200_MAX_PORT; p++)
-			if (portmap & BIT(p))
-				xrx200sw_write_x(index, XRX200_PCE_DEFPVID_PVID, p);
-	}
-}
-
 // swconfig interface
 static void xrx200_hw_init(struct xrx200_hw *hw);
 
@@ -551,7 +509,6 @@ static int xrx200sw_set_vlan_vid(struct
 	tem.val[0] = val->value.i;
 	xrx200_pce_table_entry_write(&tem);
 
-	xrx200sw_fixup_pvids();
 	return 0;
 }
 
@@ -625,9 +582,6 @@ static int xrx200sw_set_vlan_ports(struc
 
 		if (tem.val[0] == 0)
 			continue;
-
-		if ((untagged & (tem.val[1] ^ tem.val[2])) && (val->port_vlan != i))
-			return -EINVAL;
 	}
 
 	tem.index = val->port_vlan;
@@ -655,8 +609,6 @@ static int xrx200sw_set_vlan_ports(struc
 	ltq_switch_w32_mask(0, portmap, PCE_PMAP3);
 	hw->vlan_port_map[val->port_vlan] = portmap;
 
-	xrx200sw_fixup_pvids();
-
 	return 0;
 }
 
@@ -705,7 +657,6 @@ static int xrx200sw_set_vlan_enable(stru
 	tev.valid = val->value.i;
 	xrx200_pce_table_entry_write(&tev);
 
-	xrx200sw_fixup_pvids();
 	return 0;
 }
 
@@ -738,6 +689,30 @@ static int xrx200sw_get_port_pvid(struct
 	return 0;
 }
 
+static int xrx200sw_set_port_pvid(struct switch_dev *dev, int port, int val)
+{
+	int i;
+	struct xrx200_pce_table_entry tev;
+
+	if (port >= XRX200_MAX_PORT)
+		return -EINVAL;
+
+	tev.table = XRX200_PCE_ACTVLAN_IDX;
+
+	for (i = 0; i < XRX200_MAX_VLAN; i++)
+	{
+		tev.index = i;
+		xrx200_pce_table_entry_read(&tev);
+		if (tev.key[0] == val)
+		{
+			xrx200sw_write_x(i, XRX200_PCE_DEFPVID_PVID, port);
+			return 0;
+		}
+	}
+
+	return -EINVAL;
+}
+
 static int xrx200sw_get_port_link(struct switch_dev *dev,
 				  int port,
 				  struct switch_port_link *link)
@@ -879,6 +854,7 @@ static const struct switch_dev_ops xrx20
 	.get_vlan_ports = xrx200sw_get_vlan_ports,
 	.set_vlan_ports = xrx200sw_set_vlan_ports,
 	.get_port_pvid = xrx200sw_get_port_pvid,
+	.set_port_pvid = xrx200sw_set_port_pvid,
 	.reset_switch = xrx200sw_reset_switch,
 	.get_port_link = xrx200sw_get_port_link,
 //	.get_port_stats = xrx200sw_get_port_stats, //TODO
