--- a/drivers/net/ethernet/lantiq_xrx200.c
+++ b/drivers/net/ethernet/lantiq_xrx200.c
@@ -35,6 +35,7 @@
 #include "lantiq_pce.h"
 #include "lantiq_xrx200_sw.h"
 
+#define LINK_INITIAL_UP
 #define SW_POLLING
 #define SW_ROUTING
 
@@ -1359,6 +1360,11 @@ static int xrx200_mdio_probe(struct net_
 
 	phy_start_aneg(phydev);
 #endif
+
+#ifdef LINK_INITIAL_UP
+	netif_carrier_on(dev);
+#endif
+
 	return 0;
 }
 
